openshift/odo
# Proof of concept: plugin architecture based on event bus + plugin listeners, using go-plugin

## What is the purpose of this change? What does it change?
This proof of concept takes the work accomplished in #1173  and changes the underlying mechanism to use hashicorp/go-plugin instead, using plain RPC.

## Was the change discussed in an issue?
No.

## How to test changes?
- Go to a directory *outside* of your `$GOPATH` (unless you know how to deal with go modules installed in `$GOPATH`)
- `git clone https://github.com/metacosm/odo-tracer-listener.git`
- `cd odo-tracer-listener`
- `git checkout go-plugin`
- `go build -o tracer.listener.odo listener.go`
- `mkdir -p ~/.odo/plugins/` (if you don't already have that dir)
- `cp tracer.listener.odo ~/.odo/plugins/`
- Build this PR and issue an `odo` command that uses the new `oc` command pattern (any `service`, `catalog` or `link/unlink` commands, only these commands are automatically wired to dispatch events)
- You should see some plugin-related debugging messages
- Check `~/.odo/plugins/tracer.log` to see the logging performed by the tracer plugin.

Note: It might happen that the command you issued failed with an error and you might need to check that the plugin process is not still running: `ps aux | grep tracer` and then `killall tracer.listener.odo`. Basically, plugins are "managed" and need to be cleaned-up before the application exits. This is done by a deferred call in the main `odo` function. However, calls to `os.Exit` don't run deferred functions. I have added some code to perform plugins clean up before some calls to `os.Exit` but might have missed some, which might lead dangling plugin processes running (basically a server process). A proper solution to this issue will have to be devised.
